waddrmgr/address.go

```diff
@@ -402,6 +420,92 @@ func (a *managedAddress) DerivationInfo() (KeyScope, DerivationPath, bool) {
 	return a.manager.Scope(), a.derivationPath, true
 }
 
+// signature abstracts over a signature that can be verified given a message
+// and public key.
+type signature interface {
+	// Verify returns true if the sig is valid, and false otherwise.
+	Verify([]byte, *btcec.PublicKey) bool
+}
+
+// Validate takes a random message and a private key and ensures that:
+//
+//  1. The private key properly maps back to the target address.
+//
+//  2. The public key generated by the private key matches the public
+//     key of the address.
+//
+//  3. We're able to generate a valid ECDSA/Schnorr signature based on
+//     the passed private key validated against the internal public key.
+//
+// NOTE: This is part of the ValidatableManagedAddress interface .
+func (a *managedAddress) Validate(msg [32]byte, priv *btcec.PrivateKey) error {
+	// First, we'll obtain the mapping public key from the target private
+	// key. This key should match up with the public key we store
+	// internally.
+	//
+	// This can potentially catch an error in the original mapping of the
+	// private key to a public key.
+	basePubKey := priv.PubKey()
+	if !a.pubKey.IsEqual(basePubKey) {
+		return fmt.Errorf("%w: expected %v, got %v", ErrPubkeyMismatch,
+			basePubKey.SerializeUncompressed(),
+			a.pubKey.SerializeUncompressed())
+	}
+
+	// Next, we'll verify that if we take the base public key, and generate
+	// an address of the corresponding type, that matches up w/ what we've
+	// stored internally.
+	//
+	// This can potentially catch a hardware/software error when mapping
+	// the public key to a Bitcoin address.
+	addr, err := newManagedAddressWithoutPrivKey(
+		a.manager, a.derivationPath, a.pubKey, a.compressed, a.addrType,
+	)
+	if err != nil {
+		return fmt.Errorf("unable to re-create addr: %v", err)
+	}
+	if addr.address.String() != a.address.String() {
+		return fmt.Errorf("%w: expected %v, got %v", ErrAddrMismatch,
+			addr.address.String(), a.address.String())
+	}
+
+	// With the two above checks, we'll now execute our final sanity check:
+	// make sure we can generate a signature that verifies under the target
+	// public key.
+	//
+	// TODO(roasbeef): potentially run _all_ checks then see which one
+	// fails?
+	var sig signature
+
+	addrPrivKey, _ := btcec.PrivKeyFromBytes(a.privKeyCT)
```

Comment:
> I think we should use a.PrivKey() here instead, in case the private key hasn't been decrypted yet?

So in this case (at least for lnd), the wallet is always unlocked and the key decrypted. I had it using `PrivKey`, but there's a test where it makes a "raw watch only" addr w/o unlocking the wallet. In this case, a call to `PrivKey()` fail since the wallet isn't unlocked. It'll actually deadlock since it needs a mutex (had made a private func w/o a lock to test that out). 

Also this is only called from `newManagedAddress`, which _always_ have the private key available. See where `privKeyCT is set for the first time.

Reply:
Ah, I see. Makes sense given the context of the lock and how we use it. I wonder if we should export the `Validate()` method at all then? Since it makes these assumptions in the first place.